From f724c99d861fac1fe4f70c18b0384752d0f96e1e Mon Sep 17 00:00:00 2001
From: Mustafa Emre Acer <meacer@chromium.org>
Date: Fri, 07 Feb 2025 15:47:26 -0800
Subject: [PATCH] [Merge M133][Lookalikes] Disallow hostnames containing certain ligatures

On some platforms, system fonts contain ligatures that render the
Google logo. This can be used to spoof hostnames.

This CL blocks hostnames containing unsafe ligatures with an
interstitial.

(cherry picked from commit 8831e558be2ff9567efce5f1e9ea54e47d41490a)

Bug: 391788835
Change-Id: Ida694f42b2540ec61b78c7f9f6d0732a768c16c9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6227546
Reviewed-by: Joe DeBlasio <jdeblasio@chromium.org>
Commit-Queue: Mustafa Emre Acer <meacer@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1415328}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6245515
Reviewed-by: Chris Thompson <cthomp@chromium.org>
Cr-Commit-Position: refs/branch-heads/6943@{#1486}
Cr-Branched-From: 72dd0b377c099e1e0230cc7345d5a5125b46ae7d-refs/heads/main@{#1402768}
---

--- a/chrome/browser/lookalikes/lookalike_url_navigation_throttle_browsertest.cc
+++ b/chrome/browser/lookalikes/lookalike_url_navigation_throttle_browsertest.cc
@@ -339,8 +339,12 @@
       EXPECT_TRUE(interstitial->is_signed_exchange_for_testing());
     }
 
+    bool punycode_interstitial =
+        expected_event == NavigationSuggestionEvent::kFailedSpoofChecks;
     SendInterstitialCommandSync(browser,
-                                SecurityInterstitialCommand::CMD_DONT_PROCEED);
+                                SecurityInterstitialCommand::CMD_DONT_PROCEED,
+                                punycode_interstitial);
+
     EXPECT_EQ(
         expected_suggested_url,
         browser->tab_strip_model()->GetActiveWebContents()->GetVisibleURL());
@@ -515,6 +519,26 @@
   CheckInterstitialUkm({kNavigatedUrl}, "TriggeredByInitialUrl", false);
   test_helper()->CheckSafetyTipUkmCount(0);
 }
+
+// Navigate to a domain that contains an unsafe ligature.
+IN_PROC_BROWSER_TEST_F(LookalikeUrlNavigationThrottleBrowserTest,
+                       UnsafeLigature) {
+  const GURL kNavigatedUrl = GetURL("GOOGLELOGOLIGATURE.com");
+  // Ligature checks show the punycode interstitial which doesn't have a
+  // suggested URL. Its "Don't proceed" button goes back, which is the new tab
+  // page here.
+  const GURL kSuggestedURL("chrome://newtab");
+
+  base::HistogramTester histograms;
+  TestMetricsRecordedAndInterstitialShown(
+      browser(), histograms, kNavigatedUrl, kSuggestedURL,
+      NavigationSuggestionEvent::kFailedSpoofChecks);
+
+  CheckInterstitialUkm({kNavigatedUrl}, "MatchType",
+                       LookalikeUrlMatchType::kFailedSpoofChecks);
+  CheckInterstitialUkm({kNavigatedUrl}, "TriggeredByInitialUrl", false);
+  test_helper()->CheckSafetyTipUkmCount(0);
+}
 
 // Navigate to a domain that would trigger the warning, but doesn't because it
 // fails-safe when the allowlist isn't available.
--- a/components/lookalikes/core/lookalike_url_util.cc
+++ b/components/lookalikes/core/lookalike_url_util.cc
@@ -739,6 +739,27 @@
   return false;
 }
 
+// Hostnames containing these strings are considered unsafe due to ligature
+// rendering in some fonts.
+const char* kUnsafeLigatures[] = {
+    "g_logo", "o_logo", "l_logo", "e_logo",
+    // google_logo is also unsafe, but e_logo is its substring.
+    // super_g_logo is also unsafe, but g_logo is its substring.
+    "google_g", "glogoligature", "ologoligature", "llogoligature",
+    "elogoligature",
+    // googlelogoligature is also unsafe, but elogoligature is its
+    // substring
+};
+
+bool IsUnsafeLigature(const DomainInfo& domain) {
+  for (const char* unsafe_ligature : kUnsafeLigatures) {
+    if (domain.hostname.find(unsafe_ligature) != std::string::npos) {
+      return true;
+    }
+  }
+  return false;
+}
+
 }  // namespace
 
 namespace lookalikes {
@@ -1269,6 +1290,10 @@
 }
 
 bool ShouldBlockBySpoofCheckResult(const DomainInfo& navigated_domain) {
+  if (IsUnsafeLigature(navigated_domain)) {
+    return true;
+  }
+
   // Here, only a subset of spoof checks that cause an IDN to fallback to
   // punycode are configured to show an interstitial.
   switch (navigated_domain.idn_result.spoof_check_result) {
